File: eco-core/core-plugin/src/main/java/com/willfp/ecoweapons/EcoWeaponsPlugin.java
package com.willfp.ecoweapons;

import com.willfp.eco.core.EcoPlugin;
import com.willfp.eco.core.command.impl.PluginCommand;
import com.willfp.eco.core.display.DisplayModule;
import com.willfp.ecoweapons.commands.CommandEcoweapons;
import com.willfp.ecoweapons.config.EcoWeaponsJson;
import com.willfp.ecoweapons.display.WeaponsDisplay;
import com.willfp.ecoweapons.effects.Effect;
import com.willfp.ecoweapons.effects.Effects;
import com.willfp.ecoweapons.effects.util.EffectListener;
import com.willfp.ecoweapons.util.DiscoverRecipeListener;
import lombok.Getter;
import org.bukkit.event.Listener;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.List;

public class EcoWeaponsPlugin extends EcoPlugin {
    /**
     * Instance of EcoWeapons.
     */
    @Getter
    private static EcoWeaponsPlugin instance;

    /**
     * tiers.json.
     */
    @Getter
    private final EcoWeaponsJson ecoWeaponsJson;

    /**
     * Internal constructor called by bukkit on plugin load.
     */
    public EcoWeaponsPlugin() {
        super(1241, 12134, "&#ff0000");
        instance = this;

        this.ecoWeaponsJson = new EcoWeaponsJson(this);
    }

    @Override
    protected void handleEnable() {
        for (Effect effect : Effects.values()) {
            if (effect instanceof Listener) {
                this.getEventManager().registerListener((Listener) effect);
            }
        }
    }

    @Override
    protected List<Listener> loadListeners() {
        return Arrays.asList(
                new EffectListener(),
                new DiscoverRecipeListener(this)
        );
    }

    @Override
    protected List<PluginCommand> loadPluginCommands() {
        return Arrays.asList(
                new CommandEcoweapons(this)
        );
    }

    @Override
    protected @Nullable DisplayModule createDisplayModule() {
        return new WeaponsDisplay(this);
    }
}

File: eco-core/core-plugin/src/main/kotlin/com/willfp/ecoweapons/EcoWeaponsPlugin.kt
package com.willfp.ecoweapons

import com.willfp.eco.core.EcoPlugin
import com.willfp.eco.core.command.impl.PluginCommand
import com.willfp.eco.core.display.DisplayModule
import com.willfp.eco.core.integrations.IntegrationLoader
import com.willfp.ecoweapons.commands.CommandEcoweapons
import com.willfp.ecoweapons.config.EcoWeaponsYml
import com.willfp.ecoweapons.display.WeaponsDisplay
import com.willfp.ecoweapons.fuels.FuelHandler
import com.willfp.ecoweapons.util.DiscoverRecipeListener
import com.willfp.ecoweapons.weapons.WeaponListener
import com.willfp.ecoweapons.weapons.WeaponModifierListener
import com.willfp.ecoweapons.weapons.WeaponUtils
import com.willfp.ecoweapons.weapons.toSingletonList
import com.willfp.libreforge.LibReforge
import org.bukkit.event.Listener

class EcoWeaponsPlugin : EcoPlugin(1241, 12134, "&#ff0000") {
    /**
     * ecoweapons.yml.
     */
    val ecoWeaponsYml: EcoWeaponsYml

    /**
     * Internal constructor called by bukkit on plugin load.
     */
    init {
        instance = this
        ecoWeaponsYml = EcoWeaponsYml(this)
        LibReforge.init(this)
        LibReforge.registerHolderProvider { WeaponUtils.getWeaponOnPlayer(it).toSingletonList() }
    }

    override fun handleEnable() {
        LibReforge.enable(this)
    }

    override fun handleDisable() {
        LibReforge.disable(this)
    }

    override fun handleReload() {
        LibReforge.reload(this)
        FuelHandler.createRunnable(this)
    }

    override fun loadListeners(): List<Listener> {
        return listOf(
            DiscoverRecipeListener(this),
            WeaponListener(this),
            FuelHandler(),
            WeaponModifierListener(this)
        )
    }

    override fun loadIntegrationLoaders(): List<IntegrationLoader> {
        return LibReforge.getIntegrationLoaders()
    }

    override fun loadPluginCommands(): List<PluginCommand> {
        return listOf(
            CommandEcoweapons(this)
        )
    }

    override fun createDisplayModule(): DisplayModule {
        return WeaponsDisplay(this)
    }

    override fun getMinimumEcoVersion(): String {
        return "6.15.0"
    }

    companion object {
        /**
         * Instance of EcoWeapons.
         */
        @JvmStatic
        lateinit var instance: EcoWeaponsPlugin
    }
}